### PR TITLE
Add support for SWA (left, right) with FusedAttention 

### PR DESCRIPTION
**Pull request by sudhakarsingh27**
# Description
FusedAttention supports "right" side sliding window attention for some time now. This adds support for SWA (left, right) with FusedAttention backend in TE.
(changes cherry-picked from original PR: https://github.com/NVIDIA/TransformerEngine/pull/1369)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

`transformer_engine`
- `common`
    - `fused_attn`
        - `fused_attn.cpp`
            - [x] add `bottom_right_diagonal` parameter to the API
            - [x] Edit the filters to allow sliding window config to pick arbitrary seqlen fused attn backend
        - `fused_attn_f16_arbitrary_seqlen.cu`: add `bottom_right_diagonal` parameter to the API
        - `fused_attn_fp8.cu`: add `bottom_right_diagonal` parameter to the `FADescriptor_v1` API
        - `utils.h`: add `bottom_right_diagonal` parameter to `FADescriptor_v1` API

- `pytorch`
    - `transformer.py`
        - [x] plumb `bottom_right_diagonal` through the call stack: `TransformerLayer` --> `SelfAttention`/`CrossAttention`
    - `attention`
        - `dot_product_attention`
            - `backends.py`:
                - `UnfusedDotProductAttention`
                    - [x] add `bottom_right_diagonal` parameter to the `forward` API
                        - [x] why is it not used in the `forward`?
                            - `bottom_right_alignment` is being used in the Alibi call, perhaps this should be corrected
                - `FusedAttn` custom module
                    - [x] add `bottom_right_diagonal` parameter to the `forward` API
                - `FusedAttention` module
                    - [x] plumb `bottom_right_diagonal` through the call stack
            - `dot_product_attention.py`
                - `DotProductAttention`
                    - [x] Plumb `bottom_right_diagonal` through the call stack
                    - [x] Add calculation of `bottom_right_diagonal` if it's `None`
            - `utils.py`
                - `AttentionParams`
                    - [x]
                - `get_attention_backend`
                    - update sliding window filter section
                    - update attention bias filter section
        - `multi_head_attention.py`
            - [x] Add `bottom_right_diagonal` to forward API and call
            - [x] Add calculation of `bottom_right_diagonal` if it's `None`
    - `cpp_extentions`
        - `fused_attn.py`
            - [x] plumb `bottom_right_diagonal` in `fused_attn_fwd`/`fused_attn_bwd`
    - `csrc`
        - `extension`
            - `attention.cpp`
                - [x] plumb `bottom_right_diagonal` through the call stack: `fused_attn_fwd` --> `nvte_fused_attn_fwd`
                - [x] same as above for bwd
        - `extensions.h`
            - [x] add `bottom_right_diagonal` to `fused_attn_fwd` and `fused_attn_bwd` API definitions

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention/dot_product_attention/backends.py [hunk 1]
```diff
@@ -449,6 +450,7 @@ def forward(
                     actual_seqlens_q=actual_seqlens_q if "padding" in attn_mask_type else None,
                     actual_seqlens_kv=actual_seqlens_kv if "padding" in attn_mask_type else None,
                     alibi_slopes=alibi_slopes,
+                    # (This should be replaced with `bottom_right_diagonal` which is passed from the arguments)
                     bottom_right_alignment=attn_mask_type not in ["causal", "padding_causal"],
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/backends.py]: `bottom_right_alignment` should be renamed to `bottom_right_diagonal` and the plumbed through value should be used instead of calculating it here

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 2]
```diff
@@ -418,12 +420,14 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
          // 9.6: SWA (left, 0) + top-left/bottom-right diagonal + {bshd, sbhd, thd}
          (cudnn_runtime_version >= 90600 &&
           ((window_size_left == -1 && (window_size_right == -1 || window_size_right == 0)) ||
```
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn.cpp:422]: We can probably remove this line? Because it's covered by the next line?

File: tests/pytorch/attention/test_attention.py
Comment (cyanguwa) [inline tests/pytorch/attention/test_attention.py:1]: Could you add a couple of SWA tests to the CP tests as well? I think it's just a matter of replacing (left,0) with (left, right) and test them out. Thanks!

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 3]
```diff
@@ -504,18 +508,16 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
 }
 
 // NVTE fused attention FWD with packed QKV
-// DEPRECATED: This API is deprecated.
+// DEPRECATED: This API is deprecated. (Should there be a version by which this is going to be removed? @cyang)
```
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn.cpp]: I made some changes in #2272, but will see if I can make the 2.11 deadline.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 4]
```diff
@@ -254,9 +255,17 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
...
 
+      fe::DiagonalAlignment_t const &diagonal_alignment =
+          bottom_right_diagonal ? fe::DiagonalAlignment_t::BOTTOM_RIGHT
+                                : fe::DiagonalAlignment_t::TOP_LEFT;
+      sdpa_options.set_diagonal_alignment(diagonal_alignment);
       if (cudnn_runtime_version >= 90200 && window_size_left != -1) {
         sdpa_options.set_diagonal_band_left_bound(window_size_left + 1);
       }
+      if (cudnn_runtime_version >= 90600 && window_size_right != -1) {
+        // (remove comment when reviewed) Should it be `window_size_right + 1` instead?
```
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: SWA(left, right) should describe an attention window of `left + 1 + right` elements, but cuDNN understands it as `left - 1 + 1 + right` elements, so we need to add the 1 here to `left` to make all three backends (FlashAttention, FusedAttention, UnfusedDPA) equivalent in terms of the SWA operation.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 5]
```diff
@@ -781,9 +792,18 @@ void fused_attn_arbitrary_seqlen_bwd_impl(
...
+      fe::DiagonalAlignment_t const &diagonal_alignment =
+          bottom_right_diagonal ? fe::DiagonalAlignment_t::BOTTOM_RIGHT
+                                : fe::DiagonalAlignment_t::TOP_LEFT;
+      sdpa_backward_options.set_diagonal_alignment(diagonal_alignment);
+
       if (cudnn_runtime_version >= 90200 && window_size_left != -1) {
         sdpa_backward_options.set_diagonal_band_left_bound(window_size_left + 1);
       }
+      if (cudnn_runtime_version >= 90600 && window_size_right != -1) {
+        // (remove comment when reviewed) Should it be `window_size_right + 1` instead?
```
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Same as above.

File: transformer_engine/pytorch/attention/dot_product_attention/backends.py [hunk 6]
```diff
@@ -449,6 +450,7 @@ def forward(
                     actual_seqlens_q=actual_seqlens_q if "padding" in attn_mask_type else None,
                     actual_seqlens_kv=actual_seqlens_kv if "padding" in attn_mask_type else None,
                     alibi_slopes=alibi_slopes,
+                    # (This should be replaced with `bottom_right_diagonal` which is passed from the arguments)
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/backends.py]: I wonder if the logic should be something like this: `bottom_right_alignment=attn_mask_type not in ["causal", "padding_causal"] if bottom_right_alignment is None else bottom_right_alignment`.

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 7]
```diff
@@ -200,6 +200,9 @@ class AttentionParams:
         `causal_bottom_right`, `padding_causal_bottom_right`, `arbitrary`}
     window_size : Tuple[int, int], default = None
         Sliding window attention size.
+    bottom_right_diagonal: bool, default = `True`
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/utils.py]: Is the default True here?

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 8]
```diff
@@ -873,39 +878,46 @@ def _is_fa3_supported(num_heads, num_gqa_groups, head_dim_qk, head_dim_v, qkv_dt
...
-            if not FlashAttentionUtils.is_installed:
-                FlashAttentionUtils.version_required = PkgVersion("2.3")
-            elif not FlashAttentionUtils.v2_3_plus:
-                logger.debug(
-                    "Disabling FlashAttention as sliding window attention requires flash-attn 2.3+"
-                )
-                use_flash_attention_2 = False
+    # (cyang: Why is window_size is being modified but then its value ignored
+    #  in the following else block?)
+    # else:
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/utils.py]: You are right - the code under `else` should come out of `else` because we need to factor in the changes in `check_set_window_size`. Thanks - well spotted!

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 5 above)
Comment (sudhakarsingh27) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Then I wonder why the test cases are passing then, they should all fail

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 3 above)
Comment (sudhakarsingh27) [inline transformer_engine/common/fused_attn/fused_attn.cpp]:
```suggestion
// DEPRECATED: This API is deprecated.
```

File: tests/pytorch/attention/test_attention_with_cp.py [hunk 9]
```diff
@@ -148,6 +148,7 @@ def test_cp_with_flash_attention(dtype, model, qkv_format, cp_comm_type):
     ),  # MHA
     "cp_1_3": ModelConfig(2, 4096, 12, 128, attn_bias_type="post_scale_bias"),  # MHA
     "cp_1_4": ModelConfig(2, 4096, 12, 128, attn_mask_type="causal", window_size=(512, 0)),  # MHA
+    "cp_1_5": ModelConfig(2, 4096, 12, 128, attn_mask_type="causal", window_size=(512, 512)),  # MHA
```
Comment (cyanguwa) [inline tests/pytorch/attention/test_attention_with_cp.py]: Can we replace 1_4 with 1_5 to save some CI time? Same for 2_4 and 2_5.

File: tests/pytorch/utils.py [hunk 10]
```diff
@@ -353,11 +353,11 @@ def test():
...
-            available_backends, flash_attention_backend, fused_attention_backend = test()
-            if fused_attention_backend == FusedAttnBackend[backends[i]]:
-                fused_attn_backends.append(fused_attention_backend)
+
+    for i in range(3):
+        os.environ["NVTE_FUSED_ATTN_BACKEND"] = str(i)
+        _attention_backends["backend_selection_requires_update"] = True
+        available_backends, flash_attention_backend, fused_attention_backend = test()
+        if fused_attention_backend == FusedAttnBackend[backends[i]]:
+            fused_attn_backends.append(fused_attention_backend)
```
Comment (cyanguwa) [inline tests/pytorch/utils.py:362]: Is this change intended, or it's for debugging? Would this fix the printing issue we talked about before, where the unit test wouldn't print the selection logic from this for loop?

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 3 above)
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn.cpp]: Remove this comment maybe? I'm working on #2272 but don't have a clear timeline at the moment :)

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 11]
```diff
@@ -409,6 +409,8 @@ void nvte_fused_attn_bwd_qkvpacked(
  *  \param[in]     softmax_type              Attention softmax type.
  *  \param[in]     window_size_left          Sliding window size (the left half).
  *  \param[in]     window_size_right         Sliding window size (the right half).
+ *  \param[in]     bottom_right_diagonal     Whether to align sliding window and ALiBi diagonal to the bottom right corner of the softmax matrix.
+ *  \param[in]     deterministic             Whether to execute with deterministic behaviours.
```
Comment (cyanguwa) [inline transformer_engine/common/include/transformer_engine/fused_attn.h]: We probably don't need this argument here, right?

File: transformer_engine/pytorch/attention/dot_product_attention/dot_product_attention.py [hunk 12]
```diff
@@ -1081,6 +1094,15 @@ def forward(
...
             window_size = dpa_utils.check_set_window_size(attn_mask_type, window_size)
+            if bottom_right_diagonal is None:
+                bottom_right_diagonal = self.bottom_right_diagonal
+            if attn_mask_type in {"causal", "padding_causal"}:
+                bottom_right_diagonal = False
+            if bottom_right_diagonal is None or attn_mask_type in {
+                "causal_bottom_right",
+                "padding_causal_bottom_right",
+            }:
+                bottom_right_diagonal = True
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/pytorch/attention/dot_product_attention/dot_product_attention.py:1110]: Could we add some explanation in the init or fwd docstring that `bottom_right_diagonal` will be forced to `True` if the mask type has `bottom_right` in it. Just so users (and Greptile) don't get confused. It helps remove this comment from Greptile as well.

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 13]
```diff
@@ -938,13 +953,12 @@ def _is_fa3_supported(num_heads, num_gqa_groups, head_dim_qk, head_dim_v, qkv_dt
...
-        and (alibi_slopes_shape is not None or max_seqlen_q != max_seqlen_kv)
+        and (alibi_slopes_shape is not None)
     ):
         fu_core_attention_bias_type = "post_scale_bias"
         fu_core_attention_bias_requires_grad = False
-        if alibi_slopes_shape is None:
-            fu_core_attention_bias_shape = "1hss"
-        elif len(alibi_slopes_shape) == 1 and alibi_slopes_shape[0] == num_heads:
+
+        if len(alibi_slopes_shape) == 1 and alibi_slopes_shape[0] == num_heads:
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/utils.py:961]: Why is `alibi_slopes_shape is None` branch removed here?

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py:961]: The encapsulating if condition makes sure that this branch is always false

if(
...
and (alibi_slopes_shape is not None)
):

File: tests/pytorch/utils.py (hunk 10 above)
Comment (sudhakarsingh27) [reply tests/pytorch/utils.py:362]: apparently, its use here is a bit problematic and so removing it helps print the information. (I vaguely remember that you suggested me to remove it and then I checked `logging` documentation and you were right)
```
 with logging_context(highest_level=AttentionLogging._log_level):
```

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py (hunk 13 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py:961]: Ah I see. It's already included in L956.

File: tests/pytorch/attention/test_attention.py
Comment (sudhakarsingh27) [reply tests/pytorch/attention/test_attention.py:1]: Done.

File: tests/pytorch/utils.py (hunk 10 above)
Comment (sudhakarsingh27) [reply tests/pytorch/utils.py:362]: I checked that it doesn't always print the Debug information and responds to `NVTE_DEBUG*` flags so leaving this as it is here.

**Comment by cyanguwa** [comment]
You may need to add the extra argument `bottom_right_diagonal` to the `nvte_xxx` C API calls in Jax too, otherwise TE-Jax won't build. Have a look at PR 2584 maybe, where the `deterministic` arg was added, so files such as jax/csrc/extensions/attention.cpp need to be changed as well. Thanks!
